File: clients/intellij/src/main/kotlin/com/github/meymchen/lspfanalysis/model/ServerPath.kt
package com.github.meymchen.lspfanalysis.model

import java.nio.file.Path
import java.nio.file.Paths

/** Where the binary came from, which decides what to say when it is missing. */
enum class ServerSource { CONFIGURED, DEVELOPMENT, BUNDLED }

data class ResolvedServer(val binary: Path, val source: ServerSource)

/** The server binary's name, which only Windows spells differently. */
fun executableName(osName: String = System.getProperty("os.name").orEmpty()): String =
    if (osName.startsWith("Windows", ignoreCase = true)) "lspf-analysis.exe" else "lspf-analysis"

/** Everything needed to decide which binary to launch. */
data class ServerLocation(
    /**
     * The directory the packaged binary lives in, or `null` when the plugin
     * cannot work out where it was installed.
     */
    val bundledDirectory: Path?,
    /** True when running out of a Gradle sandbox rather than an installation. */
    val development: Boolean,
    /** The repository root, known only in development. */
    val repositoryRoot: Path? = null,
    /** What the language server path setting is set to, if anything. */
    val configuredPath: String? = null,
    val osName: String = System.getProperty("os.name").orEmpty(),
    val homeDirectory: String = System.getProperty("user.home").orEmpty(),
)

/**
 * Decides which binary to launch, most specific source first.
 *
 * A configured path always wins, since someone who set it means it. Failing
 * that, a sandbox IDE runs the debug build from the repository, and an
 * installed plugin runs the one packaged beside it.
 */
fun resolveServerBinary(location: ServerLocation): ResolvedServer {
    val configured = location.configuredPath?.trim()
    if (!configured.isNullOrEmpty()) {
        return ResolvedServer(
            Paths.get(expandHome(configured, location.homeDirectory)).toAbsolutePath().normalize(),
            ServerSource.CONFIGURED,
        )
    }

    val executable = executableName(location.osName)
    val repositoryRoot = location.repositoryRoot
    if (location.development && repositoryRoot != null) {
        return ResolvedServer(
            repositoryRoot.resolve("target").resolve("debug").resolve(executable).normalize(),
            ServerSource.DEVELOPMENT,
        )
    }
    // A bare name when the install directory could not be worked out: it will
    // not exist either, and the reader gets the same explanation of why.
    val bundled = location.bundledDirectory?.resolve(executable)?.normalize()
        ?: Paths.get(executable)
    return ResolvedServer(bundled, ServerSource.BUNDLED)
}

/** Expands a leading `~`, which a hand-written setting is likely to contain. */
fun expandHome(candidate: String, homeDirectory: String): String {
    if (candidate == "~") {
        return homeDirectory
    }
    if (candidate.startsWith("~/") || candidate.startsWith("~\\")) {
        return Paths.get(homeDirectory, candidate.substring(2)).toString()
    }
    return candidate
}

/**
 * Explains a binary that is not there.
 *
 * For a packaged plugin the likely cause is a ZIP built for another platform:
 * only Windows is published today, so anyone else installing it lands here, and
 * saying so is more useful than reporting a bare path.
 */
fun describeMissingServer(resolved: ResolvedServer): String {
    val found = t("server.missing", resolved.binary.toString())
    val advice = when (resolved.source) {
        ServerSource.CONFIGURED -> t("server.missing.configured")
        ServerSource.DEVELOPMENT -> t("server.missing.development")
        ServerSource.BUNDLED -> t("server.missing.bundled")
    }
    return "$found $advice"
}

File: clients/vscode/src/serverPath.ts
import * as os from 'node:os';
import * as path from 'node:path';

import { t } from './i18n.js';

/** Where the binary came from, which decides what to say when it is missing. */
export type ServerSource = 'configured' | 'development' | 'bundled';

export interface ResolvedServer {
  binary: string;
  source: ServerSource;
}

/** The server binary's name, which only Windows spells differently. */
export function executableName(platform: NodeJS.Platform): string {
  return platform === 'win32' ? 'lspf-analysis.exe' : 'lspf-analysis';
}

/**
 * What the server process is started with, short of the transport flag.
 *
 * `--stdio` is deliberately absent: `TransportKind.stdio` has
 * vscode-languageclient append it to `Executable.args` on the way to `spawn`,
 * and the server's parser refuses the flag twice — which failed the launch
 * with `the argument '--stdio' cannot be used multiple times` before the
 * client had said a word.
 */
export const SERVER_ARGS: readonly string[] = ['serve'];

/**
 * The command line that actually reaches a server spawned over stdio.
 *
 * Mirrors what vscode-languageclient adds, so the argv can be checked, and
 * run against the real binary, without an editor to host the extension.
 */
export function stdioArgs(args: readonly string[] = SERVER_ARGS): string[] {
  return [...args, '--stdio'];
}

export interface ServerLocation {
  extensionPath: string;
  /** True when running out of an Extension Development Host. */
  development: boolean;
  /** What `lspfAnalysis.server.path` is set to, if anything. */
  configuredPath?: string;
  platform?: NodeJS.Platform;
  homeDirectory?: string;
}

/**
 * Decides which binary to launch, most specific source first.
 *
 * A configured path always wins, since someone who set it means it. Failing
 * that, a development host runs the debug build from the workspace — the
 * extension lives two levels below the repository root — and an installed
 * extension runs the one packaged beside it.
 */
export function resolveServerBinary({
  extensionPath,
  development,
  configuredPath,
  platform = process.platform,
  homeDirectory = os.homedir(),
}: ServerLocation): ResolvedServer {
  const configured = configuredPath?.trim();
  if (configured) {
    return {
      binary: path.resolve(expandHome(configured, homeDirectory)),
      source: 'configured',
    };
  }
  const executable = executableName(platform);
  return development
    ? {
        binary: path.resolve(extensionPath, '..', '..', 'target', 'debug', executable),
        source: 'development',
      }
    : { binary: path.join(extensionPath, 'server', executable), source: 'bundled' };
}

/** Expands a leading `~`, which a hand-written setting is likely to contain. */
export function expandHome(candidate: string, homeDirectory: string): string {
  if (candidate === '~') {
    return homeDirectory;
  }
  if (candidate.startsWith('~/') || candidate.startsWith('~\\')) {
    return path.join(homeDirectory, candidate.slice(2));
  }
  return candidate;
}

/**
 * Explains a binary that is not there.
 *
 * For a packaged extension the likely cause is a VSIX built for another
 * architecture, which VS Code installs without complaint when it is handed
 * the file directly; saying so is more useful than reporting a bare path.
 */
export function describeMissingServer({ binary, source }: ResolvedServer): string {
  const found = t('LSPF Analysis could not find its language server at {0}.', binary);
  switch (source) {
    case 'configured':
      return `${found} ${t('Check the lspfAnalysis.server.path setting.')}`;
    case 'development':
      return `${found} ${t('Run cargo build in the repository first.')}`;
    case 'bundled':
      return `${found} ${t(
        'This usually means the installed extension was built for a different ' +
          'platform. Install the build matching this machine, or point ' +
          'lspfAnalysis.server.path at an lspf-analysis binary.',
      )}`;
  }
}
